### DISCUSSION TITLE
Langfuse batch prompts api should return prompt configs as well

### DISCUSSION THREAD

**Discussion by ashutoshsaboo** [langfuse/langfuse#2684] [Ideas]
### Describe the feature or potential improvement

Currently `GET /api/public/v2/prompts` only returns a list of prompts, but doesn't return configs with it. It becomes tough to build applications with it, as the config can contain crucial metadata. For example, if I want a entire list of prompts that apply to a user based on say their country code, and multiple other params which are included in the metadata, then I can't do that. First getting all prompts, and then doing a specific prompt detail lookup `GET /api/public/v2/prompts/{promptName}` will be too costly. 

Can this be please added? @marcklingen

**Comment by marcklingen** [comment]
The `GET /api/public/v2/prompts` by design only returns versions/labels of a specific prompt to map which prompt versions are available. As `config` varies between versions of a prompt and can be large, we did not include it in the response.

I understand that being able to filter would be really useful here, what do you think about `GET /api/public/v2/prompts/{promptName}/versions` to fetch all versions of a single prompt including the prompt and config?

**Reply by ashutoshsaboo** [reply]
So let me explain the way we are using prompts DB in langfuse, it is something like this, 

There are a bunch of topics which users can engage with in our product. All these topics are mapped 1:1 to each of their prompts in langfuse. The thing is that such super topics list is too many in number, only subset of such topics(prompts) are shown to each user based on which cohort they belong to. The targeting details, such as which country should a topic be visible in, and so on many other such targeting params are stored in the prompt config in langfuse. 

So the way we do is that, we first fetch all possible prompts from langfuse, and then we filter it based on the prompt configs to decide which prompts/topics should actually be shown to the user. 

In our prod builds, we only fetch the latest prompts whichever are with "production" label, using `/api/public/v2/prompts?label=development`
In our beta builds, we fetch the latest prompts whichever are with "development" label (a prod prompt is also in development by default for us, based on how we promote prompts), using `/api/public/v2/prompts?label=production`. 

In my opinion, for the endpoint `GET /api/public/v2/prompts` whenever a specific label as query param is passed, it would anyways only correspond to a single prompt by design in langfuse (for ex: i can't assign prod label to multiple version of same prompts in LF, similarly i also can't assign development label to multiple versions of the same prompt, this is actually a good design enforcement which simplifies things for us) -- in that case we should also return the prompt configs in the same response. For the default case, where no query label param is passed, we can keep it same as the current case. 

Regarding your suggestion of `/api/public/v2/prompts/{promptName}/versions`, as explained above, the thing is that we want to avoid multiple redundant calls for various prompts (topics in our case) to LF. It adds to significant latency, we have 100s of prompts in our LF (soon to reach in the order of 1000s), imagine making so many calls to LF - it'd be sheer catastrophic. Beside as I explained earlier given the fact that for a single label only single version corresponds, then why even waste resources, best is to return config in the same response itself! I'm also cool if you want to give the exact same functionality in a separate named endpoint, as long as the functionality is sufficed (I'm not sure why that's needed though). 

I had one more request if possible for the same `GET /api/public/v2/prompts` endpoint, for all the latest matching prompt versions that it returns based on the query label params that were passed, we should add some sort of lastUpdatedTimestamp in the final response to allow for clients to build their own caching logic on top of it, to reduce redundant calls. This can be added for the entire `GET /api/public/v2/prompts` endpoint, i.e, no need to just add for the case where label query param was passed. I'm sure each prompt version's timestamp is stored in the db anyways, we just have to return it appropriately for the correct matched latest versions based on the request params. Subsequently, we can add a query param timestamp to both the GET /prompts/ and /prompts/{promptName} apis to only return prompts that match timestamp greater than the previous one which client had already fetched for. Timestamp based caching is a very common use-case which many LF consumer clients might need in production at some point or the other (many other saas services provide this for the same reason). 

I hope that gives you sufficient context now on where I'm coming from? @marcklingen Happy to discuss and provide more details if needed! 

It's perfectly okay if lastUpdatedTimestamp comes later on (if in case it needs more changes), but I'd truly appreciate some help with the above one of returning configs in the same request.

**Reply by ashutoshsaboo** [reply]
Hi Marc, just checking if you got a chance to check this^? @marcklingen

**Reply by marcklingen** [reply]
sorry for the slow response, this was a great read and helped me understand your goals here!

lastUpdatedAt was a quick change, added it here: https://github.com/langfuse/langfuse/pull/2820

I'm yet unsure about the config, I agree that we could add it to the response especially when a filter is applied. I am just unsure whether a schema that is not stable might be confusing for some folks. Let me think about this again, will try to add to this thread later today.

Do you expect the config to be stable across versions of a prompt?

**Reply by ashutoshsaboo** [reply]
No i dont expect stable structure across versions of a prompt - config structure can change across versions, but that shouldn't be an issue right? You can just return the dict as is, and expect clients to transform it to their respective pydantic models on their end? Not sure why you are worried about the schema changing, imo  config should be just returned as a dict simply by langfuse, rest the clients can manage on their end if they wish to transform it to any model.

**Reply by marcklingen** [reply]
Sorry, this was confusing, let me be more clear. Schema changing = config is included in the response when label filter is provided but null otherwise on GET /prompts

**Reply by marcklingen** [reply]
Wdyt about returning the last (highest version nunber) config on GET /v2/prompts for each prompt (name)? Thereby the response is consistent no matter how the api is used and this should solve for your use case

**Reply by ashutoshsaboo** [reply]
> Sorry, this was confusing, let me be more clear. Schema changing = config is included in the response when label filter is provided but null otherwise on GET /prompts

I don't think this is an issue at all tbh. This can be clearly documented in langfuse api docs, that it won't return config when label filter is not provided - and I think it's the right design decision as well, many providers have such a design especially when they have query filter based apis etc. I don't think this is an "inconsistent" api output structure as well, as config would be clearly demarcated as optional in the response api docs - so for non filter based invocations it would return None, whereas for filter based invocations it would return the corresponding config. Personally, I can't think of any practical usecases as well where a user might want to query configs of all the versions of a prompt in a single api call - it doesn't make sense at all logically. So don't think this design decision should hinder any practical usecase in any way. 

> Wdyt about returning the last (highest version nunber) config on GET /v2/prompts for each prompt (name)? Thereby the response is consistent no matter how the api is used and this should solve for your use case

Even if you return the latest version, the issue would be that I would have to query config for that specific version number. Which is costly for me, because of the number of prompts that we have stored in Langfuse. I don't think this would be scalable. My usecase would require getting configs along with prompt names for the mentioned query label filter param. If you are worried about changing structure of the existing endpoint, you can always create a new endpoint which suffices for the same usecases, but still returns configs as well in the same output (if you ask me, creating a new endpoint just for something this trivial would be extremely overkill imo), rather than forcing user to call LF APIs to get config for each prompt one at a time. The latter is not great design and not scalable for clients. Imagine if a product has an extremely high DAU, and each time we are forced to query LF so many times for each prompt to get the config. The number of LF API calls would go through the roof, and cause downtime wherever LF is deployed.

**Reply by marcklingen** [reply]
Okay, I believe I have landed on a good implementation.

The `GET /v2/prompts` endpoint always returns the `lastConfig` for each prompt. If no filters are provided, it returns the last configuration for each prompt. If you add a filter (such as label or version), it returns the configuration for each prompt with the given filter.

The relevant pull request is available here: https://github.com/langfuse/langfuse/pull/2824

Could you confirm that this implementation addresses your use case? You will be able to retrieve all configurations for a given prompt label via the paginated `GET /v2/prompts` endpoint.

**Reply by marcklingen** [reply]
@ashutoshsaboo aiming to merge the PR later today or tomorrow, feel free to share any feedback you might have, hope this solves for your use case (which sounds really exciting)!

**Reply by ashutoshsaboo** [reply]
@marcklingen just replied on your linked PR. First up, thanks a bunch for all your prompt responses above. I truly appreciate it! 😄 

One additional suggestion, is it possible to add a date filter query param to the above api (which relies on the same lastUpdatedDate for each prompt)? Basically if say the client passes, `?datetime=<5th august 2024, 11:50:50 hrs (in either epoch timestamp/datetime format however you persist it in LF's Postgres)>`, then essentially LF should only return matching prompts which were last updated on or after the above timestamp. Similarly if `?datetime=xyz&label=production` is passed, then it should retrieve only the prod prompts which were last updated >= xyz timestamp whichever was passed. I.e, the datetime query filter should work with other already existing filters - label/version/both whichever the client might want to pass. I believe once this is added, this should make it feasible for almost any high-scale prod usecase to be easily built on top of LF prompts. 😄 

The idea is to cache prompts locally on the client (which we already do with our existing system, where we mange all our prompts before we move to LF), and not invoke LF unnecessarily for the already fetched prompts. Again to give you clarity - the major constraint is that, because this is going to be invoked for each user - so at extremely high DAU/concurrent calls, even the best of LF deployments can have major down-times (plus, why to really query Postgres unnecessarily when it's not even needed!). I thought of mentioning this here, rather than in the PR as I wasn't aware if you'd want this to be part of the same PR, or if you would want a separate PR for the same.

**Reply by marcklingen** [reply]
thanks, added this to the pr

**Reply by marcklingen** [reply]
merged, thanks again for your feedback here!

**Reply by ashutoshsaboo** [reply]
🚀 🚀

**Comment by ashutoshsaboo** [comment]
Hey I'll check the PR tomorrow. Thanks!

**Comment by ashutoshsaboo** [comment]
Hey @marcklingen , one follow up thing related to the above - one sort of a bit (major) miss that I see in Langfuse is that anyone can promote prompts to prod (like even non admins can promote prompts to prod). Ideally there should be an access control mechanism to only allow "admins" to promote prompts to prod. Else it can be a bit too risky. I might want to grant access to multiple people to play around with our data in Langfuse, but promoting things to prod and such critical access mechanisms should probably only be reserved for select members, be it admins or a new role. 

What do you think? Can this be added please?

**Reply by marcklingen** [reply]
This is great feedback, can you create a new idea thread for this? https://langfuse.com/ideas

We plan to add more customizable role-based access control and this seems to be a great use-case here.
